### ISSUE TITLE
[Pion-to-Pion] SACK Cumulative ACK 1837106535 is older than ACK point 1837106535

### ISSUE THREAD

**Issue by joeky888** [pions/webrtc#194]
Hi guys, I am trying to make a audio chat app using Pion-to-Pion example

Code here: https://github.com/joeky888/pions-examples/tree/master/alsa-pion-to-pion

It was fine at first, but after 2 or 3 seconds, I got a lot of error messages:

```sh
Failed to push SCTP packet: Failed handling chunk: Failure handling SACK: SACK Cumulative ACK 1837106535 is older than ACK point 1837106535                                                  
Failed to push SCTP packet: Failed handling chunk: Failure handling SACK: SACK Cumulative ACK 1837106535 is older than ACK point 1837106535                                                  
Failed to push SCTP packet: Failed handling chunk: Failure handling SACK: SACK Cumulative ACK 1837106535 is older than ACK point 1837106535                                                  
Failed to push SCTP packet: Failed handling chunk: Failure handling SACK: SACK Cumulative ACK 1837106535 is older than ACK point 1837106535                                                  
Failed to push SCTP packet: Failed handling chunk: Failure handling SACK: SACK Cumulative ACK 1837106535 is older than ACK point 1837106535                                                  
Failed to push SCTP packet: Failed handling chunk: Failure handling SACK: SACK Cumulative ACK 1837106535 is older than ACK point 1837106535                                                  
Failed to push SCTP packet: Failed handling chunk: Failure handling SACK: SACK Cumulative ACK 1837106535 is older than ACK point 1837106535                                                  
Failed to push SCTP packet: Failed handling chunk: Failure handling SACK: SACK Cumulative ACK 1837106535 is older than ACK point 1837106535                                                  
Failed to push SCTP packet: Failed handling chunk: Failure handling SACK: SACK Cumulative ACK 1837106535 is older than ACK point 1837106535                                                  
Failed to push SCTP packet: Failed handling chunk: Failure handling SACK: SACK Cumulative ACK 1837106535 is older than ACK point 1837106535                                                  
Failed to push SCTP packet: Failed handling chunk: Failure handling SACK: SACK Cumulative ACK 1837106535 is older than ACK point 1837106535                                                  
Failed to push SCTP packet: Failed handling chunk: Failure handling SACK: SACK Cumulative ACK 1837106535 is older than ACK point 1837106535                                                  
Failed to push SCTP packet: Failed handling chunk: Failure handling SACK: SACK Cumulative ACK 1837106535 is older than ACK point 1837106535                                                  
Failed to push SCTP packet: Failed handling chunk: Failure handling SACK: SACK Cumulative ACK 1837106535 is older than ACK point 1837106535                                                  
Failed to push SCTP packet: Failed handling chunk: Failure handling SACK: SACK Cumulative ACK 1837106535 is older than ACK point 1837106535                                                  
Failed to push SCTP packet: Failed handling chunk: Failure handling SACK: SACK Cumulative ACK 1837106535 is older than ACK point 1837106535                                                  
Failed to push SCTP packet: Failed handling chunk: Failure handling SACK: SACK Cumulative ACK 1837106535 is older than ACK point 1837106535                                                  
Failed to push SCTP packet: Failed handling chunk: Failure handling SACK: SACK Cumulative ACK 1837106535 is older than ACK point 1837106535
```

**Comment by Sean-Der** [comment]
Thanks for reporting this!

@kc5nra any ideas? I can check it out Friday/this weekend if not!

**Comment by joeky888** [comment]
Fyi, the code is based on my another alsa example:

https://github.com/joeky888/pions-examples/tree/master/alsa-echo

**Comment by joeky888** [comment]
And these are the changes I made:

Answer:

```diff
diff --git a/webrtc/examples/pion-to-pion/answer/main.go b/pions-examples/alsa-pion-to-pion/answer/main.go
index d58dc0e..2f468c8 100644
@@ -1,6 +1,9 @@
 package main
 
 import (
+	"syscall"
+	"os/signal"
+	"os"
 	"encoding/json"
 	"flag"
 	"fmt"
@@ -11,14 +14,51 @@ import (
 	"github.com/pions/webrtc"
 	"github.com/pions/webrtc/pkg/datachannel"
 	"github.com/pions/webrtc/pkg/ice"
+
+	"github.com/cocoonlife/goalsa"
+	"gopkg.in/hraban/opus.v2"
 )
 
+var ctrlc = make(chan os.Signal)
+
 func main() {
 	addr := flag.String("address", ":50000", "Address to host the HTTP server on.")
 	flag.Parse()
 
 	offerChan, answerChan := mustSignalViaHTTP(*addr)
 
+	const channels = 2
+	const sampleRate = 48000
+	const format = alsa.FormatS16LE
+	const pcmsize = 480
+
+	player, err := alsa.NewPlaybackDevice("default", channels, format, sampleRate,
+		alsa.BufferParams{BufferFrames: 0, PeriodFrames: pcmsize, Periods: pcmsize})
+	if err != nil {
+		panic(err)
+	}
+	defer player.Close()
+
+	recorder, err := alsa.NewCaptureDevice("default", channels, format, sampleRate,
+		alsa.BufferParams{BufferFrames: 0, PeriodFrames: pcmsize, Periods: pcmsize})
+	if err != nil {
+		panic(err)
+	}
+	defer recorder.Close()
+
+	signal.Notify(ctrlc, os.Interrupt, syscall.SIGTERM)
+	go cleanup(player, recorder)
+
+	enc, err := opus.NewEncoder(sampleRate, channels, opus.AppVoIP)
+	if err != nil {
+		panic(err)
+	}
+
+	dec, err := opus.NewDecoder(sampleRate, channels)
+	if err != nil {
+		panic(err)
+	}
+
 	/* Everything below is the pion-WebRTC API, thanks for using it! */
 
 	// Prepare the configuration
@@ -51,22 +91,40 @@ func main() {
 		d.OnOpen = func() {
 			fmt.Printf("Data channel '%s'-'%d' open. Random messages will now be sent to any connected DataChannels every 5 seconds\n", d.Label, d.ID)
 
-			for range time.NewTicker(5 * time.Second).C {
-				message := randSeq(15)
-				fmt.Printf("Sending %s \n", message)
+			pcm := make([]int16, pcmsize)
+			opusdata := make([]byte, 1000)
 
-				err := d.Send(datachannel.PayloadString{Data: []byte(message)})
+			for {
+				recorder.Read(pcm)
+
+				n, err := enc.Encode(pcm, opusdata) // pcm to opus
+				if err != nil {
+					panic(err)
+				}
+				opusdata = opusdata[:n] // Remove unused space after encoding to opus
+
+				err = d.Send(datachannel.PayloadBinary{Data: opusdata})
 				check(err)
+
+				// fmt.Printf("Sending %s \n", message)
 			}
 		}
 
 		// Register message handling
 		d.Onmessage = func(payload datachannel.Payload) {
+			pcm := make([]int16, pcmsize)
+
 			switch p := payload.(type) {
 			case *datachannel.PayloadString:
 				fmt.Printf("Message '%s' from DataChannel '%s' payload '%s'\n", p.PayloadType().String(), d.Label, string(p.Data))
 			case *datachannel.PayloadBinary:
-				fmt.Printf("Message '%s' from DataChannel '%s' payload '% 02x'\n", p.PayloadType().String(), d.Label, p.Data)
+				// fmt.Printf("Message '%s' from DataChannel '%s' payload '% 02x'\n", p.PayloadType().String(), d.Label, p.Data)
+				_, err := dec.Decode(p.Data, pcm)
+				if err != nil {
+					panic(err)
+				}
+				fmt.Println(p.Data)
+				player.Write(pcm)
 			default:
 				fmt.Printf("Message '%s' from DataChannel '%s' no payload \n", p.PayloadType().String(), d.Label)
 			}
@@ -130,3 +188,12 @@ func check(err error) {
 		panic(err)
 	}
 }
+
+func cleanup(player *alsa.PlaybackDevice, recorder *alsa.CaptureDevice) {
+	// User hit Ctrl-C, clean up
+	<-ctrlc
+	fmt.Println("Close devices")
+	player.Close()
+	recorder.Close()
+	os.Exit(1)
+}
```

**Comment by joeky888** [comment]
Offer:

```diff
diff --git a/webrtc/examples/pion-to-pion/offer/main.go b/pions-examples/alsa-pion-to-pion/offer/main.go
index 80d88c7..cd84ca3 100644
@@ -1,6 +1,9 @@
 package main
 
 import (
+	"syscall"
+	"os/signal"
+	"os"
 	"bytes"
 	"encoding/json"
 	"flag"
@@ -12,12 +15,49 @@ import (
 	"github.com/pions/webrtc"
 	"github.com/pions/webrtc/pkg/datachannel"
 	"github.com/pions/webrtc/pkg/ice"
+
+	"github.com/cocoonlife/goalsa"
+	"gopkg.in/hraban/opus.v2"
 )
 
+var ctrlc = make(chan os.Signal)
+
 func main() {
 	addr := flag.String("address", ":50000", "Address that the HTTP server is hosted on.")
 	flag.Parse()
 
+	const channels = 2
+	const sampleRate = 48000
+	const format = alsa.FormatS16LE
+	const pcmsize = 480
+
+	player, err := alsa.NewPlaybackDevice("default", channels, format, sampleRate,
+		alsa.BufferParams{BufferFrames: 0, PeriodFrames: pcmsize, Periods: pcmsize})
+	if err != nil {
+		panic(err)
+	}
+	defer player.Close()
+
+	recorder, err := alsa.NewCaptureDevice("default", channels, format, sampleRate,
+		alsa.BufferParams{BufferFrames: 0, PeriodFrames: pcmsize, Periods: pcmsize})
+	if err != nil {
+		panic(err)
+	}
+	defer recorder.Close()
+
+	signal.Notify(ctrlc, os.Interrupt, syscall.SIGTERM)
+	go cleanup(player, recorder)
+
+	enc, err := opus.NewEncoder(sampleRate, channels, opus.AppVoIP)
+	if err != nil {
+		panic(err)
+	}
+
+	dec, err := opus.NewDecoder(sampleRate, channels)
+	if err != nil {
+		panic(err)
+	}
+
 	// Prepare the configuration
 	config := webrtc.RTCConfiguration{
 		IceServers: []webrtc.RTCIceServer{
@@ -47,22 +87,40 @@ func main() {
 	dataChannel.OnOpen = func() {
 		fmt.Printf("Data channel '%s'-'%d' open. Random messages will now be sent to any connected DataChannels every 5 seconds\n", dataChannel.Label, dataChannel.ID)
 
-		for range time.NewTicker(5 * time.Second).C {
-			message := randSeq(15)
-			fmt.Printf("Sending %s \n", message)
+		pcm := make([]int16, pcmsize)
+		opusdata := make([]byte, 1000)
 
-			err := dataChannel.Send(datachannel.PayloadString{Data: []byte(message)})
+		for {
+			recorder.Read(pcm)
+
+			n, err := enc.Encode(pcm, opusdata) // pcm to opus
+			if err != nil {
+				panic(err)
+			}
+			opusdata = opusdata[:n] // Remove unused space after encoding to opus
+
+			err = dataChannel.Send(datachannel.PayloadBinary{Data: opusdata})
 			check(err)
+
+			// fmt.Printf("Sending %s \n", message)
 		}
 	}
 
 	// Register the Onmessage to handle incoming messages
 	dataChannel.Onmessage = func(payload datachannel.Payload) {
+		pcm := make([]int16, pcmsize)
+
 		switch p := payload.(type) {
 		case *datachannel.PayloadString:
 			fmt.Printf("Message '%s' from DataChannel '%s' payload '%s'\n", p.PayloadType().String(), dataChannel.Label, string(p.Data))
 		case *datachannel.PayloadBinary:
-			fmt.Printf("Message '%s' from DataChannel '%s' payload '% 02x'\n", p.PayloadType().String(), dataChannel.Label, p.Data)
+			// fmt.Printf("Message '%s' from DataChannel '%s' payload '% 02x'\n", p.PayloadType().String(), dataChannel.Label, p.Data)
+			_, err := dec.Decode(p.Data, pcm)
+			if err != nil {
+				panic(err)
+			}
+			fmt.Println(p.Data)
+			player.Write(pcm)
 		default:
 			fmt.Printf("Message '%s' from DataChannel '%s' no payload \n", p.PayloadType().String(), dataChannel.Label)
 		}
@@ -118,3 +176,12 @@ func check(err error) {
 		panic(err)
 	}
 }
+
+func cleanup(player *alsa.PlaybackDevice, recorder *alsa.CaptureDevice) {
+	// User hit Ctrl-C, clean up
+	<-ctrlc
+	fmt.Println("Close devices")
+	player.Close()
+	recorder.Close()
+	os.Exit(1)
+}
```

**Comment by kc5nra** [comment]
Will look into this, this is perfect thanks!

**Comment by Lesterpig** [comment]
I have the same issue: in my case, it is triggered by a stress test on a data channel between Firefox 63 and this package (sending 16KB packets very fast).

**Comment by schollz** [comment]
I am also having this issue, randomly at seems. The problem was alleviated a little bit by using 1kb packets instead of 32kb packets, but ends up happening eventually.

**Comment by discordianfish** [comment]
Having the same problem while trying to write a game (js, raw webrtc) with pions/webrtc as 'server'.
This worked fine locally but now that I'm running the server on a machine in the cloud, I get this error:

```
Failed to push SCTP packet: Failed handling chunk: Failure handling SACK: SACK Cumulative ACK 1182707229 is older than ACK point 1182707229
```

- Server: https://github.com/discordianfish/infisk8-server
- Client: https://github.com/discordianfish/infisk8

**Comment by schollz** [comment]
Yep, I'm getting the same kind of thing:

```
Failed to push SCTP packet: Failed handling chunk: Failure handling SACK: SACK Cumulative ACK 2991622189 is older than ACK point 2991622189
```

Sometimes it happens to the sender, sometimes it happens to the recipient. You can try my code pretty easily:

**Sender:**

```bash
$ git clone https://github.com/schollz/webrtc-transfer
$ cd webrtc-transfer/sender
$ go build -v -o sender && ./sender
```

**Recipient:**

```bash
$ git clone https://github.com/schollz/webrtc-transfer
$ cd webrtc-transfer/recipient
$ go build -v && ./recipient
```

Then paste the key from the sender to the recipient and then paste the resulting key from the recipient back into the sender. It should transfer the "sender" binary, but it seems to always get caught with the SACK problem.

**Comment by hugoArregui** [comment]
I'm experiencing the same problem between pion and js. I think it would be useful if the developers can release a a patch to at least detect the situation (in case we need to wait for the proper fix), allowing us to close the connection or something. Otherwise the pion peer dies and there's not much we can do AFAICS

**Comment by kc5nra** [comment]
I've created a repro using a packet reorder-er so I'm very close to dialing this in.  Thanks for the reports and examples!

**Comment by kc5nra** [comment]
I'll be pushing a PR that should address this, thanks for the patience!

**Comment by hugoArregui** [comment]
@kc5nra sorry to bother you, but if you have a branch with some fix, even if it's not perfect, can you share it?. This issue is fully blocking development for me.

Thanks for working on this problem!

**Comment by kc5nra** [comment]
Fixed in #302

**Comment by schollz** [comment]
Thanks @kc5nra
